src/Build/Construction/Solution/SolutionFile.cs

```diff
@@ -1211,6 +1218,50 @@ internal void ParseNestedProjects()
 
                 proj.ParentProjectGuid = parentProjectGuid;
             } while (true);
+
+            ProjectFileErrorUtilities.VerifyThrowInvalidProjectFile(ContainsNoCyclicalNesting(), "SubCategoryForSolutionParsingErrors",
+                new BuildEventFileInfo(FullPath, _currentLineNumber, 0), "SolutionContainsCyclicallyNestedProjectFiles", _solutionFile);
+        }
+
+        private bool ContainsNoCyclicalNesting()
+        {
+            SearchState[] parents = new SearchState[_projectsInOrder.Count];
+            for (int a = 0; a < _projectsInOrder.Count; a++) {
+                parents[a] = SearchState.Unchecked;
+            }
+            for (int i = 0; i < _projectsInOrder.Count; i++) {
+                if (!AncestorsContainNoCycle(parents, i))
+                {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        private bool AncestorsContainNoCycle(SearchState[] parents, int index)
```

Comment:
```suggestion
        private bool AncestorsContainNoCycle(SearchState[] state, int index)
```

? I'm finding all the indexing very confusing.